File: 02. Algorithms/09. Dynamic Programming/010. Delete and Earn/code.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        vector<int> cnt(10002);
		for(int i=0;i<nums.size();i++)
		{
			cnt[nums[i]]+=nums[i];
		}

		vector<int> dp(10002);
		dp[1]=cnt[1];
		dp[2]=max(cnt[1],cnt[2]);
		for(int i=3;i<dp.size();i++)
		{
			dp[i]=max(dp[i-2]+cnt[i],dp[i-1]);
		}

		return max(dp[10000],dp[9999]);
    }
};

File: 02. Algorithms/09. Dynamic Programming/010. Delete and Earn/code.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var deleteAndEarn = function(nums) {
    const hash = new Map();
    
    nums.forEach((num) => {
        const numCache = hash.get(num);
        if(numCache) {
            hash.set(num, numCache + 1);
        } else {
            hash.set(num, 1);
        }
    });
    
    const uniqueNums = new Set();
    
    nums.forEach((num) => {
        uniqueNums.add(num);
    });
    
    nums = [...uniqueNums];
    
    nums.sort((a,b) => {
        return a-b;
    });
    
    let earn1 = 0;
    let earn2 = 0;
    
    for(let i = 0; i < nums.length; i++) {
        const currentEarn = nums[i] * hash.get(nums[i]);
    
        if(nums[i] === nums[i-1] + 1) {
            const temp = earn2;
            earn2 = Math.max(earn2, currentEarn + earn1);
            earn1 = temp;
        } else {
            const temp = earn2;
            earn2 = currentEarn + earn2;
            earn1 = temp;
        }
    }
    
    return earn2;
    
};
